<a href="https://colab.research.google.com/github/OVP2023/E-CAP-2024/blob/main/%D0%A4%D0%BE%D1%80%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5_%D0%BF%D1%80%D0%B8%D0%B7%D0%BD%D0%B0%D0%BA%D0%BE%D0%B2_%D0%B8_%D0%BE%D0%B1%D1%83%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D0%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Формирование признаков и обучение модели

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
train = pd.read_parquet('/content/drive/MyDrive/ecap/mydata2.parquet')
train.head(7)

variantid1  variantid2  target  Cosine_similarity_names  \
0  1447875869  1447872068       1                 1.000000   
1  1176231201   284733670       1                 0.989477   
2   658617865   549848659       0                 0.966868   
3   719320625   719370486       1                 1.000000   
4  1067645658   949954740       0                 0.986480   
5   504121134   304997246       0                 0.991861   
6   690909146  1044397832       1                 0.938379   

                   art1            color1          country1         brand1  \
0                  None  [светло-бежевый]          [Япония]           None   
1                  None              None          [Россия]  [Джага Джага]   
2               [83809]              None      [Нидерланды]       [Hill's]   
3                  None     [серебристый]           [Китай]      [Elitech]   
4                  None              None              None       [Россия]   
5                  None              None  [Великобритания]     [Puriferm]   
6  [77RP008_MASTER KIT]              None              None    [MasterKit]   

                  art2            color2          country2         brand2  
0                 None  [светло-бежевый]          [Япония]           None  
1                 None              None          [Россия]  [Джага Джага]  
2              [10039]              None           [Чехия]       [Hill's]  
3                 None     [серебристый]           [Китай]      [Elitech]  
4                 None         [розовый]          [Россия]   [Нет бренда]  
5                 None              None  [Великобритания]     [Puriferm]  
6  [MASTERKIT_77RP008]              None              None    [MasterKit]

In [ ]:
train.info(0)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168516 entries, 0 to 1168515
Data columns (total 12 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   variantid1               1168516 non-null  int64  
 1   variantid2               1168516 non-null  int64  
 2   target                   1168516 non-null  int64  
 3   Cosine_similarity_names  1168516 non-null  float64
 4   art1                     289818 non-null   object 
 5   color1                   716140 non-null   object 
 6   country1                 908484 non-null   object 
 7   brand1                   902926 non-null   object 
 8   art2                     289322 non-null   object 
 9   color2                   716747 non-null   object 
 10  country2                 908664 non-null   object 
 11  brand2                   902944 non-null   object 
dtypes: float64(1), int64(3), object(8)
memory usage: 107.0+ MB


In [ ]:
train['color1'].fillna(0, inplace=True)
train['color2'].fillna(0, inplace=True)
train['brand1'].fillna(0, inplace=True)
train['brand2'].fillna(0, inplace=True)
train['country1'].fillna(0, inplace=True)
train['country2'].fillna(0, inplace=True)


for i in range(0, train.shape[0]):
    if train['art1'][i]==train['art2'][i] and train['art1'][i]!=None and train['art2'][i]!=None:
        train.loc[i,'yesart']=1
        train.loc[i,'noart']=0
    elif train['art1'][i]!=train['art2'][i] and train['art1'][i]!=None and train['art2'][i]!=None:
        train.loc[i,'noart']=1
        train.loc[i,'yesart']=0
    else:
        train.loc[i,'yesart']=0
        train.loc[i,'noart']=0

    if type(train['color1'][i])!=int and type(train['color2'][i])!=int:
        if train['color1'][i][0] == train['color2'][i][0]:
            train.loc[i,'yescolor']=1
            train.loc[i,'nocolor']=0
        else:
            train.loc[i,'nocolor']=1
            train.loc[i,'yescolor']=0
    else:
        train.loc[i,'nocolor']=0
        train.loc[i,'yescolor']=0

    if type(train['brand1'][i])!=int and type(train['brand2'][i])!=int:
        if train['brand1'][i][0] == train['brand2'][i][0]:
            train.loc[i,'yesbrand']=1
            train.loc[i,'nobrand']=0
        else:
            train.loc[i,'nobrand']=1
            train.loc[i,'yesbrand']=0
    else:
        train.loc[i,'nobrand']=0
        train.loc[i,'yesbrand']=0

    if type(train['country1'][i])!=int and type(train['country2'][i])!=int:
        if train['country1'][i][0] == train['country2'][i][0]:
            train.loc[i,'yescountry']=1
            train.loc[i,'nocountry']=0
        else:
            train.loc[i,'nocountry']=1
            train.loc[i,'yescountry']=0
    else:
        train.loc[i,'nocountry']=0
        train.loc[i,'yescountry']=0





In [ ]:
train.head(7)

variantid1  variantid2  target  Cosine_similarity_names  \
0  1447875869  1447872068       1                 1.000000   
1  1176231201   284733670       1                 0.989477   
2   658617865   549848659       0                 0.966868   
3   719320625   719370486       1                 1.000000   
4  1067645658   949954740       0                 0.986480   
5   504121134   304997246       0                 0.991861   
6   690909146  1044397832       1                 0.938379   

                   art1            color1          country1         brand1  \
0                  None  [светло-бежевый]          [Япония]              0   
1                  None                 0          [Россия]  [Джага Джага]   
2               [83809]                 0      [Нидерланды]       [Hill's]   
3                  None     [серебристый]           [Китай]      [Elitech]   
4                  None                 0                 0       [Россия]   
5                  None                 0  [Великобритания]     [Puriferm]   
6  [77RP008_MASTER KIT]                 0                 0    [MasterKit]   

                  art2            color2          country2         brand2  \
0                 None  [светло-бежевый]          [Япония]              0   
1                 None                 0          [Россия]  [Джага Джага]   
2              [10039]                 0           [Чехия]       [Hill's]   
3                 None     [серебристый]           [Китай]      [Elitech]   
4                 None         [розовый]          [Россия]   [Нет бренда]   
5                 None                 0  [Великобритания]     [Puriferm]   
6  [MASTERKIT_77RP008]                 0                 0    [MasterKit]   

   yesart  noart  yescolor  nocolor  nobrand  yesbrand  yescountry  nocountry  
0     0.0    0.0       1.0      0.0      0.0       0.0         1.0        0.0  
1     0.0    0.0       0.0      0.0      0.0       1.0         1.0        0.0  
2     0.0    1.0       0.0      0.0      0.0       1.0         0.0        1.0  
3     0.0    0.0       1.0      0.0      0.0       1.0         1.0        0.0  
4     0.0    0.0       0.0      0.0      1.0       0.0         0.0        0.0  
5     0.0    0.0       0.0      0.0      0.0       1.0         1.0        0.0  
6     0.0    1.0       0.0      0.0      0.0       1.0         0.0        0.0

In [ ]:
train = train.drop(columns=(['art1','art2','color1','color2','brand1','brand2','country1','country2']))

In [ ]:
train.head()

variantid1  variantid2  target  Cosine_similarity_names  yesart  noart  \
0  1447875869  1447872068       1                 1.000000     0.0    0.0   
1  1176231201   284733670       1                 0.989477     0.0    0.0   
2   658617865   549848659       0                 0.966868     0.0    1.0   
3   719320625   719370486       1                 1.000000     0.0    0.0   
4  1067645658   949954740       0                 0.986480     0.0    0.0   

   yescolor  nocolor  nobrand  yesbrand  yescountry  nocountry  
0       1.0      0.0      0.0       0.0         1.0        0.0  
1       0.0      0.0      0.0       1.0         1.0        0.0  
2       0.0      0.0      0.0       1.0         0.0        1.0  
3       1.0      0.0      0.0       1.0         1.0        0.0  
4       0.0      0.0      1.0       0.0         0.0        0.0

In [ ]:
train.to_parquet('/content/drive/MyDrive/ecap/mydata3.parquet')

In [ ]:
#метод стохастического градиентного спуска и оценка точности модели
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

X=train[['Cosine_similarity_names','yesart','noart','yescolor','nocolor','yesbrand','nobrand','yescountry','nocountry']] # признаки X
y=train['target'] # целевая переменная y

#Сформируйте X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

sgd_clf = SGDClassifier(random_state=42) # классификатор на основе метода стохастического градиентного спуска (англ. Stochastic Gradient Descent SGD)
sgd_clf.fit(X_train, y_train) # обучаем классификатор
# Для расчета матрицы ошибок сначала понадобится иметь набор прогнозов, чтобы их можно было сравнивать с фактическими целями
y_train_pred = cross_val_predict(sgd_clf, X_train, y_train, cv=3)
print(confusion_matrix(y_train, y_train_pred))
print(accuracy_score(y_train, y_train_pred))

[[297804 126751]
 [148376 245030]]
0.6636428876193363


In [ ]:
from sklearn.pipeline import make_pipeline  # используем пайплайны для удобства
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

model = make_pipeline(
    StandardScaler(),
    LogisticRegression(max_iter=1000) # максимальное количество иттерация для обучения модели
)
X=train[['Cosine_similarity_names','yesart','noart','yescolor','nocolor','yesbrand','nobrand','yescountry','nocountry']] # признаки X
y=train['target'] # целевая переменная y
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
model.fit(X_train, y_train)
predictions = model.predict(X_test)

print(model.score(X_train, y_train) )
print(model.score(X_test,y_test))
print(accuracy_score(y_test, predictions))


0.6990308828905045
0.6991313773872859
0.6991313773872859


1) Пока правильность всего 55% на 2 признаках (name и артикул разбиваем на 2), нужно добавить еще 4 Признаки бренда, страны-изготовителя, типа товара, цвета
2) добавлением признака цвета увеличил до 60% 3) добавлением признака бренда увеличил до 65% 4)но добавлением признака страны-изготовителя улучшил модель только на 1.4% до 66.4%

Сравню с логистической регрессией у нее  70% получилось

In [ ]:
# Save Model Using Pickle

import pickle
#import joblib

# save the model to disk
filename = '/content/drive/MyDrive/ecap/model2.pkl'
pickle.dump(model, open(filename, 'wb'))
#joblib.dump(model, open(filename, 'wb'))


In [ ]:
# load the model from disk  это должно быть в другой программе)
loaded_model = pickle.load(open(filename, 'rb'))
result = loaded_model.score(X_test, y_test)
print(result)

0.6991313773872859
